src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocations.Analyzer.cs

```diff
@@ -0,0 +1,267 @@
+    ///     - The invocation method is not receiving a ct argument, and...
+    ///     - The invocation method either:
+    ///         - Has no overloads but its current signature receives an optional ct=default, currently implicit, or...
+    ///         - Has a method overload with the exact same arguments in the same order, plus one ct parameter at the end.
+    ///         
+    /// Conditions for negative cases:
+    ///     - The containing method signature does not receive a ct parameter.
+    ///     - The invocation method signature receives a ct and one is already being explicitly passed, or...
+    ///     - The invocation method does not have an overload with the exact same arguments that also receives a ct, or...
+    ///     - The invocation method only has overloads that receive more than one ct.
+    /// </summary>
+    public abstract class ForwardCancellationTokenToInvocationsAnalyzer : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA2016";
+
+        protected abstract SyntaxNode? GetMethodNameNode(SyntaxNode invocationNode);
+
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(
+            nameof(MicrosoftNetCoreAnalyzersResources.ForwardCancellationTokenToInvocationsDescription),
+            MicrosoftNetCoreAnalyzersResources.ResourceManager,
+            typeof(MicrosoftNetCoreAnalyzersResources)
+        );
+
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(
+            nameof(MicrosoftNetCoreAnalyzersResources.ForwardCancellationTokenToInvocationsMessage),
+            MicrosoftNetCoreAnalyzersResources.ResourceManager,
+            typeof(MicrosoftNetCoreAnalyzersResources)
+        );
+
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(
+            nameof(MicrosoftNetCoreAnalyzersResources.ForwardCancellationTokenToInvocationsTitle),
+            MicrosoftNetCoreAnalyzersResources.ResourceManager,
+            typeof(MicrosoftNetCoreAnalyzersResources)
+        );
+        internal static DiagnosticDescriptor ForwardCancellationTokenToInvocationsRule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Reliability,
+            RuleLevel.BuildWarning,
+            s_localizableDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false
+        );
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(ForwardCancellationTokenToInvocationsRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(AnalyzeCompilationStart);
```

Comment:
I don't know if we care about the perf impact of passing method groups over lambdas.

```suggestion
            context.RegisterCompilationStartAction(context => AnalyzeCompilationStart(context));
```

Reply:
I suppose we care about all perf improvements. I can make the change. Do you know how this helps improve performance?